packages/flutter/lib/src/cupertino/text_selection.dart

```diff
@@ -236,8 +266,35 @@ class _CupertinoTextSelectionControls extends TextSelectionControls {
 
   /// Builder for iOS-style copy/paste text selection toolbar.
   @override
-  Widget buildToolbar(BuildContext context, Rect globalEditableRegion, Offset position, TextSelectionDelegate delegate) {
+  Widget buildToolbar(
+    BuildContext context,
+    Rect globalEditableRegion,
+    Offset position,
+    List<TextSelectionPoint> endpoints,
+    TextSelectionDelegate delegate,
+  ) {
     assert(debugCheckHasMediaQuery(context));
+
+    // The toolbar should appear below the TextField
+    // when there is not enough space above the TextField to show it.
+    final double availableHeight
+        = globalEditableRegion.top - MediaQuery.of(context).padding.top - _kToolbarScreenPadding;
+    final _ArrowDirection direction = (availableHeight > _kToolbarHeight)
+        ? _ArrowDirection.down
+        : _ArrowDirection.up;
+
+    final TextSelectionPoint startTextSelectionPoint = endpoints[0];
+    final TextSelectionPoint endTextSelectionPoint = (endpoints.length > 1)
+        ? endpoints[1]
+        : null;
+    final double x = (endTextSelectionPoint == null)
+        ? startTextSelectionPoint.point.dx
+        : (startTextSelectionPoint.point.dx + endTextSelectionPoint.point.dx) / 2.0;
+    final double y = (direction == _ArrowDirection.up)
+        ? startTextSelectionPoint.point.dy + globalEditableRegion.height + _kToolbarHeight
+        : startTextSelectionPoint.point.dy - globalEditableRegion.height;
+    final Offset position = Offset(x, y);
```

Comment:
I wouldn't redeclare `position` since it's also passed in.  Maybe change the name, or at least reassign it like `position = Offset(x, y);`.  Though I usually also try not to reassign or redeclare parameters.

Reply:
Yes, I would try to avoid mutating parameters that are passed into the function (or shadow it).